### PR TITLE
clean up algorithmprovider pkg and remove scheduler deprecated algorithm-provider flag

### LINKED ISSUE #102151: Deprecate AlgorithmProvider flag and algorithmProvider scheduler pkg
#### What would you like to be added:

1. Remove [AlgorithmProvider](https://github.com/kubernetes/kubernetes/blob/8c0d06aad87047235beafd14546ff8418499b17e/cmd/kube-scheduler/app/options/deprecated.go#L50) flag. This flag was marked as deprecated since v1.17, we can remove it now.
2. Remove the [algorithmprovider](https://github.com/kubernetes/kubernetes/tree/master/pkg/scheduler/algorithmprovider) pkg. The default set of plugins would move to be set in the component config defaulting logic.

#### Why is this needed:
1. To enable versioning the default set of scheduler plugins 
2. Cleanup: unify defaulting for scheduler configuration 

/sig scheduling

### PR DESCRIPTION
**Pull request by Haleygo**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
1. Remove AlgorithmProvider flag in scheduler. This flag was marked as deprecated since v1.17.
2. clean up algorithmprovider pkg,  only preserve default plugin setting

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/102151

#### Does this PR introduce a user-facing change?

```release-note
The deprecated flag --algorithm-provider has been removed from kube-scheduler. Use instead ComponentConfig to configure the set of enabled plugins
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/algorithmprovider/registry.go [hunk 1]
```diff
@@ -1,178 +0,0 @@
-/*
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/registry.go:1]: Keep this pkg for now, only remove the flag. The pkg will be removed in the same PR that adds the set of default plugins to the defaulting logic.

File: pkg/scheduler/algorithmprovider/registry.go [hunk 2]
```diff
@@ -1,178 +0,0 @@
...
-	return Registry{
-		schedulerapi.SchedulerDefaultProviderName: defaultConfig,
-		ClusterAutoscalerProvider:                 caConfig,
-	}
-}
-
-// ListAlgorithmProviders lists registered algorithm providers.
-func ListAlgorithmProviders() string {
-	return fmt.Sprintf("%s | %s", ClusterAutoscalerProvider, schedulerapi.SchedulerDefaultProviderName)
-}
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/registry.go:69]: remove this.

File: pkg/scheduler/algorithmprovider/registry.go [hunk 3]
```diff
@@ -1,178 +0,0 @@
...
-			},
-		},
-	}
-	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeCapacityPriority) {
-		plugins.Score.Enabled = append(plugins.Score.Enabled, schedulerapi.Plugin{Name: volumebinding.Name, Weight: 1})
-	}
-	return plugins
-}
-
-func getClusterAutoscalerConfig() *schedulerapi.Plugins {
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/registry.go:157]: remove this function

File: pkg/scheduler/algorithmprovider/registry.go [hunk 4]
```diff
@@ -1,178 +0,0 @@
...
-		Bind: schedulerapi.PluginSet{
-			Enabled: []schedulerapi.Plugin{
-				{Name: defaultbinder.Name},
-			},
-		},
-	}
-	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeCapacityPriority) {
-		plugins.Score.Enabled = append(plugins.Score.Enabled, schedulerapi.Plugin{Name: volumebinding.Name, Weight: 1})
-	}
-	return plugins
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/registry.go:154]: call `applyFeatureGates(plugins)`

File: pkg/scheduler/algorithmprovider/registry.go [hunk 5]
```diff
@@ -1,178 +0,0 @@
...
-		ClusterAutoscalerProvider:                 caConfig,
-	}
-}
-
-// ListAlgorithmProviders lists registered algorithm providers.
-func ListAlgorithmProviders() string {
-	return fmt.Sprintf("%s | %s", ClusterAutoscalerProvider, schedulerapi.SchedulerDefaultProviderName)
-}
-
-func getDefaultConfig() *schedulerapi.Plugins {
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/registry.go:71]: export this function

File: pkg/scheduler/factory.go [hunk 6]
```diff
@@ -20,6 +20,23 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
```
Comment (ahg-g) [inline pkg/scheduler/factory.go:23]: as I mentioned in the other comment, please revert all changes here, keep the code where it is, only remove the flag.

File: pkg/scheduler/algorithmprovider/registry.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/algorithmprovider/registry.go:1]: /cc @yuzhiquan

**Comment by Haleygo** [comment]
@ahg-g I already revert all the changes about algorithmprovider pkg and modify the description of this pr, can you have a second review, thank you

**Comment by ahg-g** [comment]
> @ahg-g I have already reverted all the changes about algorithmprovider pkg and modify the description of this pr, can you have a second review, thanks a lot

sorry, may be I wasn't clear, but I did leave some comments on the deleted file indicating what we should keep. In [algorithmprovider/registry.go](https://github.com/kubernetes/kubernetes/blob/781382a74db02368887f726e4633e58d62f1e3ab/pkg/scheduler/algorithmprovider/registry.go#L53) please remove `Registry`, `NewRegistry`, `ClusterAutoscalerProvider`, `ListAlgorithmProviders` and `getClusterAutoscalerConfig`. Export `getDefaultConfig` and at the end of `getDefaultConfig` `applyFeatureGates`.

**Comment by ahg-g** [comment]
please update the release note as follows: `The deprecated flag --algorithm-provider has been removed from kube-scheduler. Use instead ComponentConfig to configure the set of enabled plugins`

**Comment by Haleygo** [comment]
> > @ahg-g I have already reverted all the changes about algorithmprovider pkg and modify the description of this pr, can you have a second review, thanks a lot
> 
> sorry, may be I wasn't clear, but I did leave some comments on the deleted file indicating what we should keep. In [algorithmprovider/registry.go](https://github.com/kubernetes/kubernetes/blob/781382a74db02368887f726e4633e58d62f1e3ab/pkg/scheduler/algorithmprovider/registry.go#L53) please remove `Registry`, `NewRegistry`, `ClusterAutoscalerProvider`, `ListAlgorithmProviders` and `getClusterAutoscalerConfig`. Export `getDefaultConfig` and at the end of `getDefaultConfig` call `applyFeatureGates(plugins)` before returning.

my bad, all fixed now

File: pkg/scheduler/algorithmprovider/registry.go [hunk 7]
```diff
@@ -151,18 +124,10 @@ func getDefaultConfig() *schedulerapi.Plugins {
 	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeCapacityPriority) {
```
Comment (ahg-g) [inline pkg/scheduler/algorithmprovider/registry.go:124]: can you please move this to `applyFeatureGates` func below please.

Comment (Haleygo) [reply pkg/scheduler/algorithmprovider/registry.go]: Of course, done

**Comment by ahg-g** [comment]
/assign @liggitt 

This is the flag we discussed removing offline. Can you PTAL?

**Comment by ahg-g** [comment]
/lgtm 

from the scheduler

**Comment by ahg-g** [comment]
@Haleygo please squash the commits.

**Comment by Haleygo** [comment]
> /lgtm
> 
> for the scheduler

@ahg-g So sorry, I thought remove flag and clean up pkg should divided into two commits and I have rebased commits now, but the /lgtm label been removed.

**Comment by liggitt** [comment]
> This is the flag we discussed removing offline. Can you PTAL?

looking today

**Comment by liggitt** [comment]
> please update the release note as follows: `The deprecated flag --algorithm-provider has been removed from kube-scheduler. Use instead ComponentConfig to configure the set of enabled plugins`

A release note that points to the specific documentation someone moving away from that flag should use would be helpful. It looks like https://kubernetes.io/docs/reference/scheduling/config/ is the page that should link to. It would also be helpful if there was a section in the docs specifically about how to use component config to do equivalent/similar things to what the algorithm provider / policy used to allow you to do.

**Comment by roycaihw** [comment]
The deprecation looks good to me. Could you update the release note to mention the website documentation about the replacement of the flag? https://kubernetes.io/docs/reference/scheduling/config/

Ideally the documentation should mention how should the ComponentConfig look like if a system admin is using the deprecated algorithm-provider flag is certain ways today.

**Comment by pacoxu** [comment]
/retest
volume flake？

**Comment by pacoxu** [comment]
Flake: MetricsGrabber should grab all metrics from a Scheduler
/retest

**Pull request merged**